### ISSUE TITLE
Number of iterations

### ISSUE THREAD

**Issue by jaewoosong** [amplab/SparkNet#55]
Good afternoon!

Thank you for your kind answers about my previous questions. I really appreciate your help.

May I ask about the number of iterations? I ran `CifarApp.scala` with two different settings: `cifar10_full` and `cifar10_quick`. The variable `max_iter` is set to `60000` and `4000` in the `solver.prototxt` files, respectively.

But when I ran `CifarApp.scala`, it continued learning to more than 206100 iterations. So I cancelled the job and read the source code. It seemed that the only way to quit from `while(true)` in `CifarApp.scala` is the failure of `assert` statements in the infinite loop.

May I ask how to control the number of iterations? Is the value of `max_iter` in `solver.prototxt` files are ignored in `CifarApp.scala`? It seems that when prototxt files are loaded, the parameters will go through some transformation by `ProtoLoader.replaceDataLayers`, `loadSolverPrototxtWithNet` and `CaffeNet()` but I am not sure whether I understand the code correctly.

If I am asking a basic question, please forgive me. Thank you so much for your invaluable help!

**Comment by pcmoritz** [comment]
Excellent point, we were thinking that most people would probably implement their own Apps for fitting, and CifarApp is supposed to be just an example, so we did not implement the proper stopping condition. If you insert code like

```scala
    val solverParameter = ProtoLoader.loadSolverPrototxt(sparkNetHome + "/caffe/examples/cifar10/cifar10_full_solver.prototxt");
    val maxIter = solverParameter.getMaxIter();
```
before the while loop, you can replace the while loop by another while loop that stops at the right iteration.

But you are right it would be good if the default CifarApp would already have this behavior. Let me know if you would like to submit a pull request for this.

**Comment by jaewoosong** [comment]
Hi! Thank you for your answer.

I'm having a hard time understanding how many iterations are in one while loop. May I ask the relationship between the number of iterations and the number of while loops in `CifarApp.scala`? For example, with `cifar10_quick_solver.prototxt`, the number of iterations on the screen is 80 times the `var i` in the loop. `i` is 275 when the number of iteration is 22000.

Meanwhile, I think I should adjust the number of iterations by `maxIter / (numTrainMinibatches / trainBatchSize)` but I'm not sure. May I also ask about this?

Always thank you for your help!

**Comment by robertnishihara** [comment]
The line `workerStore.getNet("net").train(syncInterval)` does training for `syncInterval` iterations (each iteration is one minibatch update). In the case of `CifarApp`, we set `syncInterval = 10`. Each training minibatch has size `trainBatchSize`, which is 100 in this app. This means that each time `i` increments by 1, the net processes 10 minibatches (or equivalently 1000 images) on each worker.

If you add a stopping condition, it should probably check how many minibatch updates the net has done. That is, it should probably check something like `i < maxIters / syncInterval`.

I'm not sure where the number 80 comes from. Where is the number of iterations being logged (are they being logged by Caffe or by SparkNet)? And how did you obtain them?

Also, is the behavior the same when you use the default prototxt files in `CifarApp.scala`?

**Comment by jaewoosong** [comment]
Thank you for your kind explanation. And sorry for my late reply.

If I understood correctly, the exact iteration checking criterion would be like below. Would you please check whether my code is right?

```scala
val solverParameter = ProtoLoader.loadSolverPrototxt(sparkNetHome +
                      "/caffe/examples/cifar10/cifar10_quick_solver.prototxt")
val maxIter = solverParamater.getMaxIter()
val originalTrainBatchSize = trainTestParameter.getTrainBatchSize()
                             // assumed this function exists.
val maxLoop = (maxIter * originalTrainBatchSize) / (syncInterval * trainBatchSize)
while(i < maxLoop) { ... }
```

About the 80-time-relationship between `i` and the number of iterations, `i` was from the log files from SparkNet (e.g., `training_log_1455174284730.txt`) and the number of iterations were from the screen output from Caffe (e.g., `I0211 15:04:58.478055  3542 solver.cpp:236] Iteration 143, loss = 1.95104`).

Now I guess that maybe due to the delay for writing log to files from memory buffer. When iteration goes larger, then the relationship is close to 80. For example, iteration 22000 at `i = 275`. But when iteration is smaller, then the relationship is only loosely 80, like iteration 500 at `i = 6` and iteration 1000 at `i = 12`. I tested on both `cifar10_quick_solver.prototxt` and `cifar10_full_solver.prototxt`.

Always thank you for your help. Happy new year! (It's lunar new year here!)

**Comment by robertnishihara** [comment]
The training batch size used in the code is called `trainBatchSize`. What is `originalTrainBatchSize`? Is that the training batch size specified in the prototxt file? If so, we probably want them to be the same.

**Comment by jaewoosong** [comment]
Hi! Thank you for your reply.

Yes, `originalTrainBatchSize` was an imaginary variable intended to mean the training batch size specified in the prototxt file. I wrote that variable because I was not sure whether the training batch size would be automatically synced with `trainBatchSize` or not.

I figured out the '80' relationship! Iteration on the screen was `i * (syncInterval * number of workers)`. Last time I used 8 workers for both `cifar10_quick` and `cifar10_full`, and that was why the iteration on the screen was 80 times `i`. I tried 1, 5 and 8 workers and the relationship still remained. When there was only one worker, iteration was 10 times `i`. With 5 workers, iteration was 50 times `i`.

May I ask the number of workers will change the training result? If `i` is set to be `i < maxIters / syncInterval`, the iteration on the screen will differ depending on the number of workers. For example, 1000 for 1 worker and 8000 for 8 workers.

I read the SparkNet paper and I remember that distributed training results are combined later. So is 1000 training iterations with 1 worker is equal to 8000 training iterations with 8 workers? I guess 1 iteration (1 worker) is divided into 8 sub-iterations (8 workers), and those 8 sub-iterations will be combined to make 1 iteration, but I'm not sure. If my guess is true, then the loop condition is `i < maxIters  / syncInterval`. But if my guess is false, that is if 1 iteration with 1 worker is same as 1 iteration with 8 workers, then I think the loop condition will be `i < maxIters / (syncInterval * number of workers)`.

Always thank you for your kind reply. Have a good day!

**Comment by robertnishihara** [comment]
Ah, sorry about the slow response. The behavior you're seeing is a little surprising to me. Each worker has its own instantiation of Caffe, so if you have 5 workers, they should all be logging separately, and their iterations shouldn't add together.

Are you running SparkNet on a cluster? Or are you running a Spark cluster locally?

We changed the code base a bunch today. It should be more stable now. Let me know if you still see the same behavior with the new code!

**Comment by jaewoosong** [comment]
Thank you for your reply! Always thank you for your response.

I am running SparkNet on my local machine. That might be the reason that the iteration number on the screen equals to `i * (syncInterval * number of workers)`. The number of iterations will not add together on clusters but might add together if SparkNet is run on a local machine.

If that is the case, then I think the while loop condition should be `i < maxIters / syncInterval` because 1 iteration of 1-worker setting will be same as 8 iterations of 8-worker setting (in the local mode).

I was just looking at the new code when you wrote me a reply! I always surprise from your great amount of work.

I tried the newest revision. On the local mode setting, this time the iteration on the screen is always 0:
```
I0217 17:12:22.015837  1120 sgd_solver.cpp:106] Iteration 0, lr = 0.001
entering forwardBackward
transformInto took 0.005 s
ForwardBackward took 0.244 s
I0217 17:12:22.265413  1120 sgd_solver.cpp:106] Iteration 0, lr = 0.001
entering forwardBackward
transformInto took 0.005 s
ForwardBackward took 0.244 s
```

In the training log file, iterations are numbered normally:
```
192.28, i = 30: broadcasting weights
192.282, i = 30: setting weights on workers
192.293, i = 30: testing
205.313, i = 30: 42.07% accuracy
205.313, i = 30: training
208.065, i = 30: collecting weights
208.076, i = 30: weight = -0.0051024556
208.076, i = 31: broadcasting weights
208.08, i = 31: setting weights on workers
208.091, i = 31: training
211.051, i = 31: collecting weights
211.061, i = 31: weight = -0.0046465653
```

Thank you for your help!

**Comment by jaewoosong** [comment]
Hi! May I ask how you think about the loop termination condition again?

In the local mode with `val maxIter = solverParameter.getMaxIter();`, I think the loop termination condition should be `while ( i < syncInterval * number of workers )`.

If there is one worker node and `syncInterval = 10`, then for each iteration `i`, 10 minibatches will be processed. Then if there are 8 worker nodes and `syncInterval = 10`, will 80 minibatches be processed for each iteration `i`? (Because `workerStore.getNet("net").train(syncInterval)` is applied to every individual worker.)

Then, for example in Cifar10_quick, to achieve 4000 iterations set in Caffe's original setting file, `i < 50` will be nice for 8 workers and `i < 400` will be good for 1 worker. In other words, I think the number of iterations shown on the Spark master's screen is equivalent to the number of iterations written in the Caffe setting file.
